packages/peregrine/lib/hooks/useDelayedTransition.js

```diff
@@ -0,0 +1,125 @@
+import { useEffect, useRef } from 'react';
+import { matchPath } from 'react-router';
+import { useHistory, useLocation } from 'react-router-dom';
+import { useApolloClient } from '@apollo/client';
+
+import { availableRoutes } from '@magento/venia-ui/lib/components/Routes/routes';
+import { useAppContext } from '@magento/peregrine/lib/context/app';
+import { useRootComponents } from '@magento/peregrine/lib/context/rootComponents';
+import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
+import DEFAULT_OPERATIONS from '../talons/MagentoRoute/magentoRoute.gql';
+import { getRootComponent } from '../talons/MagentoRoute/helpers';
+
+const DELAY_MESSAGE_PREFIX = 'DELAY:';
+
+const useDelayedTransition = () => {
+    const { pathname } = useLocation();
+    const history = useHistory();
+    const client = useApolloClient();
+    const operations = mergeOperations(DEFAULT_OPERATIONS);
+    const { resolveUrlQuery } = operations;
+    const [, setComponentMap] = useRootComponents();
+    const [, appApi] = useAppContext();
+    const { actions: appActions } = appApi;
+    const { setPageLoading } = appActions;
+    const unblock = useRef();
+
+    useEffect(() => {
+        // Override globalThis.addEventListener to prevent binding beforeunload while we add our blocker
```

Comment:
Can you explain more about this situation? I don't love the idea of overriding `addEventListener`, so I'd like to understand what the issue is and why this solves it.

Reply:
@jimbo In react-router's `history.block`, they add a new event listener for the beforeunload to pop the default window.confirm to block navigation. This would mean that when you reload the page or navigate externally we'd get the message. With this we're blocking the event binding just for our block.

Reply:
@jimbo https://github.com/remix-run/history/blob/main/docs/blocking-transitions.md#caveats

Reply:
Cool, thanks for clarifying.